### ISSUE TITLE
Yeelight LED color 10W maximum brightness

### ISSUE THREAD

**Issue by iChUdY** [vieira/homebridge-yeelight-wifi#22]
Hello,
i have Yeelight LED color 10W and install this homebridge plugin.

Maximum brightness in Home App on color temperature in not same maximum brightness in Yeelight app. In Home App brigthness is approximately 80% versus Yeelight App. This case is only for color temperature. Others color works great.

My debug work in Node-Red and Yeelight plugin found one thing. Yeelight use RGB msg for color light and CT msg(color temperature) for color tempetature light. 

So.. Home App  and this homebridge plugin use only RGB msg for both color and color temperature and result is not same like Yeelight App.

Could this plugin also use CT msg for color temperature? thanks

![6bf88719-ee07-4ee7-abfc-1a59396326eb](https://user-images.githubusercontent.com/11601525/41864769-549b8824-78ab-11e8-8a31-6db8cdce1e1f.png)
![391bc199-5d70-4b86-88d4-5875da297c5a](https://user-images.githubusercontent.com/11601525/41864770-54d27762-78ab-11e8-80f8-059659b0ad05.png)

**Comment by vieira** [comment]
Hello @iChUdY,

Yes I will have to look at CT because of #17 as Yeelight now ships some devices that are white only but support adjusting the temperature.

**Comment by iChUdY** [comment]
Thanks for react. I am willing to help you develop and test.

**Comment by iChUdY** [comment]
Any progress ?

**Comment by vieira** [comment]
Hello @iChUdY,

Sorry for the delay. Initial support is now on #25. I will still have to perform some more tests but it should mostly work.

Can you have a look and let me know how it works for you?

Thanks!

**Comment by poiromaniax** [comment]
I am also having the brightness issue. On Yeelight app, 100% brightness is much more than Home app.

Is there a way to set max brightness for all yeelight devices in config? Not 1 by 1?

**Comment by vieira** [comment]
This issue was caused by a Yeelight firmware update, multiple reports [here](https://forum.yeelight.com/t/topic/7393) and affects the bulb when in color mode (e.g. `set_hsv`). Brightness increases only to about 50% and from 50% to 100% no visible change in brightness can be seen. This is happening at the firmware level and has nothing to do with homekit or with the official app specifically.

The problem does not occur when the bulb is in color temperature mode. If a color temperature is set using `set_ct_abx`, brightness does increase correctly, however only tones of orange to bright white are supported (~1700 to 6500K).

Unfortunately, Homekit does not support light bulbs that expose both color and color temperature: 

> 8.107 Color Temperature: 
> (...)
> If this characteristic is included in the Lightbulb (page 217), Hue (page 151) and Saturation (page 159) must not be included as optional characteristics in Lightbulb (page 217).
> This characteristic must not be used for lamps which support color.

When a lamp supports color, Homekit sends the color temperature in terms of hue and saturation, that we can set using the yeelight `set_hsv` method, which has the brightness problem discussed above since the firmware update.

To sum up, this is something that used to work without problems, was broken by Yeelight in a firmware update (likely 1.4.2_0069) and should be fixed by them.

If brightness is more important than colors, I will release a new version where you can change your configuration to blacklist color. In that case we will expose color temperature (if supported) and use the `set_ct_abx` method which does not have the brightness limitation.

```json
        "defaultValue": {
            "aed78s": {
                "name": "Kitchen",
                "blacklist": ["set_hsv"]
            }
        }
```

**Comment by AdamFalconieri** [comment]
Hello, **I still have the problem with the brightness that does not increase beyond ~50%**. You didn't propose a patch associated with this topic? It would help me a lot.

Thank you in advance!

**Comment by vieira** [comment]
@AdamFalconieri Unfortunately that is a problem with the firmware since an update around December 2018, see, e.g. https://forum.yeelight.com/t/topic/8594. There is nothing I can do about it. 😞

**Comment by RealDyllon** [comment]
On my Color V2 Bulb, both the "Color" and "Temperature" settings in the Home app reduce the brightness to 50%. I need to use the Yeelight app to increase the brightness back to 100%. I haven't updated my bulb firmware yet.

If I don't use the "Color" and "Temperature" settings in the Home app, the brightness slider works flawlessly.

**Comment by RealDyllon** [comment]
@vieira I blacklisted set_hsv for my bulbID in the config.json and the problem is resolved. Brightness is back to normal (1% - 100%).

**However, the color temperature wheel broken in the Home App.** There is no "cool-daylight-blue" color present on the wheel, _see image1 attached_. 

On top of that, after selecting a temperature on the temperature wheel, **the temperature will slowly change** (temperature reduces towards cool-daylight) without me touching the wheel. This continues for as long as I am in the Color Temperature screen. 

My initial yellow shade turns into a harsh blueish light. **The logs also show the color temperature changing over time**, _see image2 attached_.  Note that I have only changed the temperature one time at the beginning, the subsequent changes to the temperature are not made by me.

image1:
![IMG_0853](https://user-images.githubusercontent.com/25902208/56849324-78a07980-6925-11e9-8954-0ca7b7577710.PNG)

image2:
![homebridge color wheel log](https://user-images.githubusercontent.com/25902208/56849329-8eae3a00-6925-11e9-9179-c6c340db6a34.png)

My Yeelight is a color v2 bulb. Model no: YLDP05YL.

**Comment by vieira** [comment]
Hello @RealDyllon,

The color wheel is not broken: it is reflecting the color temperature range supported by your bulb. Before blacklisting `set_hsv` you were able to pick a larger range of color temperatures because Homekit would use the bulb color leds to simulate color temperatures outside the supported range. Once you blacklist it, the color wheel shows a smaller range because that is what the lamp supports.

Regarding the slowing change color temperature, I will try to take a look in the weekend to reproduce the issue and hopefully implement a fix. Thanks for letting me know.

**Comment by ezeklusman** [comment]
Sorry if I'm wrong but can you post the needed updates to make it work with 100% brightness instead of color? I need that much more than to set a color.
Thanks!

**Comment by vieira** [comment]
@ezeklusman Hello, this feature is already in the latest release.

To use it, place the following snippet in your homebridge configuration (remember to replace `aed78s` in the example below with the id of your lamp):

```
{
    "platforms": [{
        "platform": "yeelight",
        "defaultValue": {
            "aed78s": {
                "name": "Kitchen",
                "blacklist": ["set_hsv"]
            }
        }
    }]
}
```

**Comment by okonet** [comment]
@vieira I removed the code that would disallow that since it seems it's supported now as of iOS 11.3. 

> Unfortunately, Homekit does not support light bulbs that expose both color and color temperature:

Can you link to the HomeKit quote?

**Comment by vieira** [comment]
> Can you link to the HomeKit quote?

I am not sure if the document is available online as a page I can link to, but it's a PDF called "HomeKit Accessory Protocol Specification" and a non-commercial version can be obtained for free via the Apple Developer portal. 

The relevant quote is on page 214, section 8.107 Color Temperature:

<img width="974" alt="Screen Shot 2019-12-19 at 23 08 52" src="https://user-images.githubusercontent.com/207460/71217136-9c4d7100-22b4-11ea-8197-37ab9ee79a19.png">

**Comment by vieira** [comment]
I just downloaded the latest version to check if something changed. It's now in page 167, section 9.21 but it seems to state the same:

<img width="914" alt="Screen Shot 2019-12-19 at 23 12 29" src="https://user-images.githubusercontent.com/207460/71217289-2a295c00-22b5-11ea-9514-464a3575bbe2.png">

**Comment by RealDyllon** [comment]
I think @okonet may be referring to the new color wheel tab switch in iOS 13.
![3E84DC3E-A978-493D-9432-430E93568BBB](https://user-images.githubusercontent.com/25902208/71224732-380cca80-2311-11ea-903d-3689c817ef4d.jpeg)
@vieira 
The new UI seems to give the impression that we can switch between the color modes as will.

**Comment by RealDyllon** [comment]
@vieira I looked into the way that hue and lifx seem to work and apparently, when the color temperature tab is selected, the bulb receives kelvin values from homekit insteas of hsv. Can we try to reproduce that here? As of now, with yeelight the color temperature tab still sends hsv values, which is the root cause of this.

**Comment by RealDyllon** [comment]
The hue and lifx bulbs still receive hsv from homekit, but I think the bulb's own firmware may be converting the hsv to ct (kelvin or otherwise) and thus displays a kelvin color. Could we try to emulate that behavior? I'll take a look at the code later and see what I can do.

**Comment by vieira** [comment]
@RealDyllon something might have changed, but last time I checked if you declared both characteristics the devices would stop appearing in Home. I went looking into the documentation to understand why and I found that paragraph.

These two modes already showed up at least since iOS 10:

![img_1988](https://user-images.githubusercontent.com/207460/71225553-e93f4080-22d0-11ea-85f2-9e16b38795ac.png)

if the device as the hsv characteristic, HomeKit will send the color temperature as a color instead of a temperature. What we receive in homebridge is just a color, no indication if the user picked that color from the color wheel or the temperature wheel. If the device supports only temperature, then it sends a proper temperature instead of a color.

I experimented with an idea which was to do some math and check if the color received was within a range that could be represented by a temperature (basically reverse what Home app must be doing), however it proved more difficult than I expected and I never finished it. This is something that should be handled by the firmware. It used to work much better with older versions.

**Comment by RealDyllon** [comment]
@vieira you're correct, we can only expose one characteristic to homekit.

**Comment by RealDyllon** [comment]
@vieira  Got a lead: https://community.lifx.com/t/expose-colour-temperature-to-homekit/3788/3

This homebridge plugin for lifx does what i was thinking. It converts from the **mirek scale on homekit** to **kelvin scale on lifx** on the homebridge plugin
https://github.com/devbobo/homebridge-lifx-lan/commit/e60ea91bc02db841af37803211b03ee4aa29287c

**Comment by okonet** [comment]
As of the latest version, both characteristics are exposed to HomeKit and it works “fine” with the yeelight Star floor lamp for me. The only issue I’m seeing is the display one but when I switch tabs I see how it affects luminosity if the lamp (temperature results in a brighter light). But the color samples Home app displayed are wrong. I thought it’s related to some kid of conversion this plugin we’re doing but now I see it can be the fact they don’t support it or something else. 

If someone can poke this around that would be cool.

**Comment by okonet** [comment]
I just noticed that [homebridge-hue did the same](https://github.com/ebaauw/homebridge-hue/releases/tag/v0.11.45). And the issue I was referring is well described here: https://github.com/ebaauw/homebridge-hue/issues/571

**Comment by vieira** [comment]
Interesting, I did not know of that change in iOS 13. That is nice as it at least now allows max brightness to be inline with what can be done with the yeelight app. It probably breaks compat with older versions of iOS/macOS, maybe that’s why the spec was not changed.

**Comment by skve** [comment]
Is there any updates to this?

**Comment by okonet** [comment]
@lukeshiels no one is working on this issue afaik so feel free to pick it up!

**Comment by incline02** [comment]
Has anyone come with an update to yeelight plugin?

or is it that the lifx plugin or hue do the work?

**Comment by vieira** [comment]
From what I can see this is working now, probably after the changes from @okonet.

In 2.4.0 I pushed a fix for updating the Home app with the correct "color" after setting a given color temperature, but the light was already changing to color temperature mode and back correctly.
